### ISSUE TITLE
espeak crashes with sum strings

### ISSUE THREAD

**Issue by alex19EP** [espeak-ng/espeak-ng#824]
i can crash espeak with sum long strange strings. 

# platforms

* nvda 20.3
* android 10

i attaching file with such strings:
[bad-strings.txt](https://github.com/espeak-ng/espeak-ng/files/5361108/bad-strings.txt)

such strings is common on webpages so for me espeak-ng is crashes  often enough.

**Comment by sukiletxe** [comment]
Cannot reproduce. Please tell us your eSpeak and NVDA settings (language, voice, variant).

**Comment by alex19EP** [comment]
> 
> 
> Cannot reproduce. Please tell us your eSpeak and NVDA settings (language, voice, variant).

sorry. not tested much crash is reproducing on Russian language. English is not affected. i am using Max variant.

**Comment by sukiletxe** [comment]
Confirmed, eSpeak doesn't say anything on these lines. Seems other languages are affected too, for example Arabic. It also seems that, in some languages, the results are different based on NVDA's punctuation levels.

**Comment by jaacoppi** [comment]
I can reproduce this with current HEAD, but not with 1.50 installed on my machine. I don't have time to bisect at the moment.

**Comment by alex19EP** [comment]
> I can reproduce this with current HEAD, but not with 1.50 installed on my machine. I don't have time to bisect at the moment.

I've done bisect first bad commit is: f0c345eb6a5d54e88e4a1aafcc9fcf2e0afa3388.

**Comment by jaacoppi** [comment]
Thanks, that's my commit. I'll create a test to make sure this doesn't happen again and fix it.

**Comment by alex19EP** [comment]
@Qchristensen fyi. nvda is affected by this too.

**Comment by beqabeqa473** [comment]
i can confirm crash on first and last strings in nvda latest with espeak-ng set to georgian language.

**Comment by jaacoppi** [comment]
@rhdunn could you help with the test? I tried adding it to `tests/language-phonemes.test` but it doesn't work. The output is the same before and after git revert f0c345e. What's the correct way to test this?

**Comment by rhdunn** [comment]
I can't reproduce it on Debian linux with the default build flags. Given that it is reproducible from the NVDA builds (using Microsoft's Visual C++ compiler), I suggest trying a build of espeak with the various build hardening options enabled (like stack protection). What also makes me think it is something like that (a stack corruption bug) is that the strings are long, and don't contain spaces. If that does turn out to be the case, it would be useful to enable those flags in the automated builds as well.

The `translate.test` file would be a logical place for this kind of thing -- the same for #831.

**Comment by valdisvi** [comment]
@jaacoppi, assuming that #831 is related issue, I [created test](https://github.com/valdisvi/espeak-ng/commit/8ef5b968eefacd569e6488d45499a1f9e66477fd) which currently fails due to segmentation fault:

```
make check
...
testing hi "---लच्छेद"
Segmentation fault
1d0
< l@c#:'e:d
make: *** [Makefile:2654: tests/translate.check] Error 1
```
Because it fails, I didn't want to add it into master branch, but put it separately as branch [831](https://github.com/valdisvi/espeak-ng/tree/831) in my project. You can use this test for your experiments.

**Comment by alex19EP** [comment]
> I can't reproduce it on Debian linux with the default build flags. Given that it is reproducible from the NVDA builds (using Microsoft's Visual C++ compiler), I suggest trying a build of espeak with the various build hardening options enabled (like stack protection). What also makes me think it is something like that (a stack corruption bug) is that the strings are long, and don't contain spaces. If that does turn out to be the case, it would be useful to enable those flags in the automated builds as well.

i experience crash with 
LDFLAGS=-Wl,-z,relro
CFLAGS=-g -O2 -fdebug-prefix-map=/root=. -fstack-protector-strong -Wformat -Werror=format-security

i think that what debian uses.

**Comment by alex19EP** [comment]
hay @Qchristensen i think NVDA 2020.4 should incorporate this fix.

**Comment by Qchristensen** [comment]
Thanks Alex,

We might be a bit late for 2020.4, but we are working on more regular updates of new eSpeak-NG releases into NVDA, so I'll flag this for 2021.1.

**Comment by jbowler** [comment]
The OP reported five strings as causing a problem.  Presumably this was on Windows because the attached file (strings.txt) was in DOS format.  As for this moment I can repro the problem with three of the strings, numbers 1, 3 and 5.  The repro does not depend on OS or compiler (as discussed elsewhere).
